### ISSUE TITLE
tctl resources

### ISSUE THREAD

**Issue by kontsevoy** [gravitational/teleport#1137]
## Problem

Teleport 2.2 has allowed for configuration and resources to merge. 2.3 needs to have a clear separation between them.

A configuration is a singleton: it always exists and there's always one copy. Not zero, not 1+.

A resource should have a true CRUD semantics: i.e. get, list, create, delete, update. We also have strange resources like SAMLResponse. What is that? Why do they exist? This is just a temporary data structure inside Teleport. Things like "reverse tunnels" or "certificate authorities" are the same: they are not user-visible resources, they are internal data structures.

## Proposal

Since we're moving configuration to a [separate issue](https://github.com/gravitational/teleport/issues/1098), the following objects need to have first-class CLI resource support:

1. Trusted Clusters
2. UserRoles (enterprise only)
3. Auth connectors like OIDC or SAML (enterprise only)

Teleport also supports low-level undocumented resources which we use for debugging or from Telekube, like "cert authorities" or "reverse tunnels". They should remain in code but we are not documenting them and keeping the right of changing their format as we wish.

## YAML format

We already have YAML format defined for the 3 resources above. 
Should we change anything there?

#### Create & Update

```
$ tctl create <flags> resources.yaml
Flags:
    -f, --force    Overwrites a resource if it already exists. (update)
```

Note: the YAML file may contain a list of resources of different types (`---` syntax).

#### Delete

```
$ tctl rm <flags> <args>
Flags:
    -f, --force    Do not return errors if a resource is not found

Args can take two forms:
   - `resource.yaml` a path to an existing file.
   - `type name1 name2 name3` resource type followed by a resource names

Examples:
  $ tctl rm role admin
  $ tctl rm admin.yaml
```

Note: the YAML file may contain a list of resources of different types (`---` syntax).

#### List & Get

```
$ tctl get <args>
Args be one of:
    - When missing, show help
    - Resource type: `tctl get roles`
    - Resource type + name: `tctl get role admin`
    - `all` a special "resource type" which means to dump all resources

Examples:
  $ tctl get roles
  $ tctl get all
  $ tctl get role admin  
  $ tctl get role admin ops
```

Notes: 
  * `tctl get all > file.yaml` should produce YAML file which then can be fed into `tctl create -f file.yaml`.
  * we do not support things like `tctl get roles users` (multiple resource types as arguments).

## Deprecate old commands

`tctl` also supports commands like `users`  and `tokens`. They are also resources, of course. Lets add a deprecation warning to them, saying that they will be removed in future versions.

We don't have to implement users and tokens resources  _right now_ to shrink the scope of this release.

## Undocumented resources

`tctl` currently supports all kinds of resources (low level) which we use for Telekube and other internal purposes. It's also a good test bed to add new resources and trying them out before making them official. The following needs to happen here:

* `tctl create` should refuse to accept resources not supported by Teleport distribution (i.e. OSS version should not accept Enterprise resources)
* `tctl create` should print a clear `DANGER` message when undocumented resource is being created.

## Other notes

* Lets not support `--format` flag for now and always default to YAML

**Comment by klizhentas** [comment]
@kontsevoy there is a missing `apply` or `upsert` that I necessary to make scripts easier - without those the scripts would look like this:

```
if tctl get resource then;
   tctl update -f resources.yaml
else
   tctl create -f resources.yaml
fi
```

This makes ansible scripts hard, with `apply` or `upsert` ansible scripts will be much simpler to write.

```bash
tctl apply -f resources.yaml
```

**Comment by kontsevoy** [comment]
@klizhentas no, you just use `tctl create -f` without any conditions. That's what `-f` (force) is for. You're probably thinking it means "file" like in Kubernetes but it means "force".

**Comment by klizhentas** [comment]
@kontsevoy question on rm command

How would one distinguish that `resource.myresource` is a filename on disk or resource name

   - `resource.yaml` a path to an existing file.
   - `type name1 name2 name3` resource type followed by a resource names

**Comment by kontsevoy** [comment]
@klizhentas re: `rm` command params:

We have two choices. The command can be dumb and "pedantic", thus requiring an additional "tell me how to interpret my own args" flag (something like `--file`) or it could apply a bit of intelligence, placing fewer restrictions to a user.

I always vote for CLI commands that aren't dumb. In this case, we can use the following information to avoid adding another flag. For the first argument to be a type instead of a file, _all of the following_ must be true:

1. It needs to belong to a known list of reserved words (allowed resource types).
2. There needs to be more than one arg (i.e. followed by the resource names)

Additionally, for an argument to be interpreted as a file, the following must be true:

1. It needs to exist
2. It must have an extension (and only `.yaml` or `.yml` are supported at this time).

The requirement to have an extension IMO is a sensible one because it leaves us an easy way to add additional resource formats in the future (JSON, text)

To summarize, the flag processing would go like:

1. Try to interpret arguments as type + names (using steps above)
2. If failed, try to interpret arguments as files.
3. If failed, show an error.

**Comment by klizhentas** [comment]
Also, our current resources are specified by `kind` and not type, so CLI should be consistent with this naming scheme

**Comment by klizhentas** [comment]
For get we need a format flag, on how to output the result - e.g. --format=json or --format=yaml for export purposes

**Comment by klizhentas** [comment]
Also optional `--with-secrets` that is set to off to allow connectors export

**Comment by kontsevoy** [comment]
Re: format. the ticket explicitly states that we're not officially supporting anything other than YAML to save time.

Not following what `--with-secrets` means.
